module/data/actor/npc.mjs

```diff
@@ -123,12 +124,24 @@ export default class NPCData extends CreatureTemplate {
   /** @inheritdoc */
   static _migrateData(source) {
     super._migrateData(source);
+    NPCData.#migrateSource(source);
     NPCData.#migrateTypeData(source);
     AttributesFields._migrateInitiative(source.attributes);
   }
 
   /* -------------------------------------------- */
 
+  /**
+   * Convert source string into custom object.
+   * @param {object} source  The candidate source data from which the model will be constructed.
+   */
+  static #migrateSource(source) {
+    if ( !source.details?.source || foundry.utils.getType(source.details.source) === "Object" ) return;
+    source.details.source = { custom: source.details.source };
```

Comment:
I think we need to be careful and ensure that `source.details.source` is present:
```suggestion
    if ( ("source" in source.details) && (foundry.utils.getType(source.details.source) !== "Object") ) {
      source.details.source = {custom: source.details.source};
    } 
```